pkg/ee/applications/util_test.go

```diff
@@ -0,0 +1,197 @@
+					MajorMinorVersion: "1.30",
+				},
+			},
+		},
+		{
+			name:      "case 2: fetching template data should fail when cluster cannot be fetched",
+			namespace: clusterNamespace,
+			seedClient: fake.
+				NewClientBuilder().WithObjects().Build(),
+			want:    nil,
+			wantErr: ErrNoClusterForTemplating,
+		},
+		{
+			name:      "case 3: fetching template data should fail when cluster has no valid version",
+			namespace: "test",
+			seedClient: fake.
+				NewClientBuilder().WithObjects(&kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-cluster",
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					HumanReadableName: "humanreadable-cluster-name",
+				},
+				Status: kubermaticv1.ClusterStatus{
+					UserEmail: "owner@email.com",
+					Address: kubermaticv1.ClusterAddress{
+						URL:  "https://cluster.seed.kkp",
+						Port: 6443,
+					},
+					Versions: kubermaticv1.ClusterVersionsStatus{
+						ControlPlane: semver.Semver("vinvalid"),
+					},
+				},
+			}).Build(),
+			want:    nil,
+			wantErr: ErrNoClusterForTemplating,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := GetTemplateData(context.TODO(), tt.seedClient, tt.namespace)
+			if !errors.Is(err, tt.wantErr) {
+				t.Fatalf("getTemplateData() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Fatalf("getTemplateData() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestRenderValueTemplate(t *testing.T) {
+	tests := []struct {
+		name         string
+		values       map[string]any
+		templateData TemplateData
+		want         map[string]any
+		wantErr      error
+	}{
+		{
+			name: "case 1: rendering for cluster name and version should succeed",
+			values: map[string]any{
+				"key1": "value1",
+				"key2": "{{ .Cluster.Name }}",
+				"key3": map[string]any{
+					"nestedkey": "{{ .Cluster.Version }}",
+				},
+			},
+			templateData: TemplateData{
+				Cluster: ClusterData{
+					Name:    "test",
+					Version: "9.9.9",
+				},
+			},
+			want: map[string]any{
+				"key1": "value1",
+				"key2": "test",
+				"key3": map[string]any{
+					"nestedkey": "9.9.9",
+				},
+			},
+			wantErr: nil,
+		},
+		{
+			name: "case 2: parsing unknown variables should lead to an error",
+			values: map[string]any{
+				"key1": "value1",
+				"key2": "{{ .Foo.Name }}",
+				"key3": map[string]any{
+					"nestedkey": "{{ .Foo.Version }}",
+				},
+			},
+			templateData: TemplateData{
+				Cluster: ClusterData{
+					Name:    "test",
+					Version: "9.9.9",
+				},
+			},
+			want:    nil,
+			wantErr: fmt.Errorf("failed to render template: "),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := RenderValueTemplate(tt.values, &tt.templateData)
+			if (err != nil) && !strings.Contains(err.Error(), tt.wantErr.Error()) {
+				t.Fatalf("getTemplateData() error = %v, wantErr %v", err, tt.wantErr)
+			}
```

Comment:
(sidenote; personally I am not a big fan of asserting error strings. It's brittle and can cause confusion when errors are reworded. If distinguishing between errors is crucial, errors should be dedicated types instead; in 99.9% of cases I personally in tests treat errors as quasi booleans that either happen or not)